File: frontend/src/services/closing.ts
import api from '../utils/api';
import type { MoneyValue } from './dashboard';

export interface ClosingAccount {
  id: string;
  name: string;
  balance: number;
}

export interface ClosingCategoryBudget {
  name: string;
  budget: number;
  actual: number;
}

export interface ClosingBudgetSummary {
  total_budget: number;
  total_spent: number;
  categories: ClosingCategoryBudget[];
}

export interface ClosingGoalProgress {
  name: string;
  progress: number;
}

export interface ClosingSnapshot {
  month: string;
  accounts: ClosingAccount[];
  total_income: number;
  total_expense: number;
  total_assets: number;
  total_debts: number;
  net_worth: number;
  total_cash: number;
  dti_ratio: number;
  health_score: number;
  ef_coverage_months: number;
  budget_summary: ClosingBudgetSummary;
  goals_progress: ClosingGoalProgress[];
}

export interface DeltaValue {
  absolute_change: number;
  formatted_absolute_change: string;
  percentage_change: number;
  direction: 'up' | 'down' | 'flat';
}

export interface ClosingComparison {
  prev_month: string;
  net_worth_delta: DeltaValue;
  assets_delta: DeltaValue;
  debts_delta: DeltaValue;
  cash_delta: DeltaValue;
  income_delta: DeltaValue;
  expense_delta: DeltaValue;
}

export interface MonthlyClosing {
  id: string;
  month: string;
  snapshot: ClosingSnapshot;
  total_income: MoneyValue;
  total_expense: MoneyValue;
  net_worth: MoneyValue;
  total_assets: MoneyValue;
  total_debts: MoneyValue;
  total_cash: MoneyValue;
  dti_ratio: number;
  ef_coverage_months: number;
  is_confirmed: boolean;
  confirmed_at?: string;
  notes: string;
  comparison?: ClosingComparison;
}

export const closingService = {
  async generateClosing(month: string, notes?: string): Promise<MonthlyClosing> {
    const res = await api.post('/monthly-closing/generate', { month, notes });
    return res.data.data;
  },

  async getClosings(): Promise<MonthlyClosing[]> {
    const res = await api.get('/monthly-closing');
    return res.data.data || [];
  },

  async getClosingDetail(month: string): Promise<MonthlyClosing> {
    const res = await api.get(`/monthly-closing/${month}`);
    return res.data.data;
  },
};
export default closingService;

File: backend/internal/dto/closing.go
package dto

type MonthlyClosingRequest struct {
	Month string `json:"month" binding:"required"`
	Notes string `json:"notes"`
}

type ClosingAccount struct {
	ID      string  `json:"id"`
	Name    string  `json:"name"`
	Balance float64 `json:"balance"`
}

type ClosingCategoryBudget struct {
	Name   string  `json:"name"`
	Budget float64 `json:"budget"`
	Actual float64 `json:"actual"`
}

type ClosingBudgetSummary struct {
	TotalBudget float64                 `json:"total_budget"`
	TotalSpent  float64                 `json:"total_spent"`
	Categories  []ClosingCategoryBudget `json:"categories"`
}

type ClosingGoalProgress struct {
	Name     string  `json:"name"`
	Progress float64 `json:"progress"`
}

type ClosingSnapshot struct {
	Month            string                `json:"month"`
	Accounts         []ClosingAccount      `json:"accounts"`
	TotalIncome      float64               `json:"total_income"`
	TotalExpense     float64               `json:"total_expense"`
	TotalAssets      float64               `json:"total_assets"`
	TotalDebts       float64               `json:"total_debts"`
	NetWorth         float64               `json:"net_worth"`
	TotalCash        float64               `json:"total_cash"`
	DTIRatio         float64               `json:"dti_ratio"`
	HealthScore      int                   `json:"health_score"`
	EFCoverageMonths float64               `json:"ef_coverage_months"`
	BudgetSummary    ClosingBudgetSummary  `json:"budget_summary"`
	GoalsProgress    []ClosingGoalProgress `json:"goals_progress"`
}

type DeltaValue struct {
	AbsoluteChange          float64 `json:"absolute_change"`
	FormattedAbsoluteChange string  `json:"formatted_absolute_change"`
	PercentageChange        float64 `json:"percentage_change"`
	Direction               string  `json:"direction"` // up, down, flat
}

type ClosingComparison struct {
	PrevMonth     string     `json:"prev_month"`
	NetWorthDelta DeltaValue `json:"net_worth_delta"`
	AssetsDelta   DeltaValue `json:"assets_delta"`
	DebtsDelta    DeltaValue `json:"debts_delta"`
	CashDelta     DeltaValue `json:"cash_delta"`
	IncomeDelta   DeltaValue `json:"income_delta"`
	ExpenseDelta  DeltaValue `json:"expense_delta"`
}

type DataSufficiency struct {
	IsSufficient       bool     `json:"is_sufficient"`
	MissingFields      []string `json:"missing_fields,omitempty"`
	UsesFallbackValues bool     `json:"uses_fallback_values"`
}

type MonthlyClosingResponse struct {
	ID               string             `json:"id"`
	Month            string             `json:"month"`
	Snapshot         ClosingSnapshot    `json:"snapshot"`
	TotalIncome      MoneyValue         `json:"total_income"`
	TotalExpense     MoneyValue         `json:"total_expense"`
	NetWorth         MoneyValue         `json:"net_worth"`
	TotalAssets      MoneyValue         `json:"total_assets"`
	TotalDebts       MoneyValue         `json:"total_debts"`
	TotalCash        MoneyValue         `json:"total_cash"`
	DTIRatio         float64            `json:"dti_ratio"`
	EFCoverageMonths float64            `json:"ef_coverage_months"`
	IsConfirmed      bool               `json:"is_confirmed"`
	ConfirmedAt      string             `json:"confirmed_at,omitempty"`
	Notes            string             `json:"notes"`
	Comparison       *ClosingComparison `json:"comparison,omitempty"`
	DataSufficiency  *DataSufficiency   `json:"data_sufficiency,omitempty"`
}
